### DISCUSSION TITLE
How can we improve PR comments?

### DISCUSSION THREAD

**Discussion by tim775** [infracost/infracost#1016] [Ideas & Feature Requests]
## Background
Inspired by #809, I've been working adding a new `post_condition={ "update": true}` option for all GitHub integrations (GitHub Actions, Azure DevOps with GitHub repos, CircleCI with GitHub repos).  This new option reduces noise by updating the infracost comment if it already exists, instead of creating a new one every time.  

We're experimenting with github discussions, so I started one so we can figure out what other improvements we can make!

## Current Behavior
With #980 the options for `post_condition` will become:

- `'{"update": true}'`: "when a commit results in a change in costs vs earlier commits, create or update a pr comment." PR followers will only be notified on the comment create, and the comment will stay at the same location in the PR comment history. This is initially for GitHub, we'll investigate how we can support it for others later.
- `'{"has_diff": true}'`: "only post a comment if there is a diff. This is the default behavior." A commit comment is put on the first commit with a terraform change, and on every subsequent commit (regardless of whether or not there is a tf change in the particular commit).
- `'{"always": true}'`: "always post a comment." A commit comment is put on every commit.
- `'{"percentage_threshold": 0}'`: "absolute percentage threshold that triggers a comment. For example, set to 1 to post a comment if the cost estimate changes by more than plus or minus 1%." A commit comment is put on every commit with a terraform change that results in a cost diff > x%

## What's next?
In thinking about what the future behavior could be it seems like the "has_diff" option isn't very useful, and that it could be useful to combine "percentage_threshold" with the other options.  I was thinking `post_condition` could be `{ "behavior": <update|replace|always>, "percentage_threshold":<int>}`:

- `'{"behavior": "update"}'`: "when a commit results in a change in costs vs earlier commits, create or update a pr comment." PR followers will only be notified on the comment create, and the comment will stay at the same location in the pr comment history.
- `'{"behavior": "replace"}'`: "when a commit results in a change in costs vs earlier commits, delete any existing pr comments and create a new one." PR followers will only be notified on each comment create, and the new comment will be at the bottom of the comment history.
- `'{"behavior": "always"}'`: "when a commit results in a change in costs vs earlier commits, minimize any existing pr comments and create a new one." PR followers will be notified on each comment create, and the new comment will be at the bottom of te comment history.

- `'{"percentage_threshold": 0}'`: "Modifies the above behaviors so that no comments are made until a commit on the pr has a cost change absolute percentage above this threshold. For example, setting to 1 will prevent comments until the cost estimate changes by more than plus or minus 1%." A commit comment is put on every commit with a terraform change that results in a cost diff > x% (regardless of whether or not there is a tf change in the particular commit )" Once you have a commit that passes the threshold, the update/replace/new behavior kicks in for the life of the pr. 

## Over to you
We'd love feedback or thoughts from anyone!

**Comment by alikhajeh1** [comment]
### Combining multiple runs into 1 PR comment

I'm collecting feedback about this sub-topic here from 3 users so far:

1. @soulshake made a great point in https://github.com/infracost/infracost/issues/809#issuecomment-863143927 as well as the follow-up comments on the need to combine multiple run outputs into 1 PR comment.

2. An Atlantis user mentioned something similar: "Our Terraform is set up into multiple repositories and most times multiple plans are run in a single PR (for each region over each Terraform repo). Each has their own Infracost estimate (which is reasonable); I expect combining these would require a lot more 'magic' but would be very nice."

3. Another user who uses TravicCI, wrote some python code to run Infracost against their dev, staging and production envs, merge the cost diffs and post 1 PR comment.

4. https://github.com/infracost/infracost-gh-action/issues/31 mentioned they run a matrix build and don't want individual comments.

5. Another user who uses hundreds of state files in their repo mentioned that: "We run states in parallel depending on which files were modified in a PR, regardless of how many states are modified or created in a PR, we want only a single comment to post to the PR."
They've copied and modified our GitHub Action so it allows them to:

    ```
    1. Generate a matrix build based on the number of modified states
    2. Run infracost against each of these states independently, and post the results as GitHub actions artifacts
    3. Post matrix / parallel build, run a GHA step that pulls down all the artifacts, aggregates them into a single report via “infracost output” (formerly “infracost report”)
    4. Post this comment to GitHub on the PR
    5. Subsequent builds will also review the comments on the PR and if there is already one from the infracost action, update it, not post it, or delete it, depending on what the result of the most recent run is. 
    Everything above allows us to create a nice single report informing how much the PR will cost, with a single comment on the PR that gets updated accordingly as things change in the PR based on feedback and subsequent commits. 
    ```

I'm not sure how we could do that without a server-side part, maybe using CI artifacts can help, it's something to think about too.

**Comment by dgcaron** [comment]
We also have several terraform deployments in our (mono) repo and I think it is quite common as it is advised to reduce/isolate the blast radius of the terraform deployments.  merging several outputs into one message would at the very top of my feature wish list. 

with regards to the update|append|percentage,  i feel that percentage is a slippery slope. when the costs go up, the percentage get's harder to hit and substantial increases in costs can still occur without people noticing. i do feel that repeating (almost) the same cost analysis over and over again will get to noisy. for a PR I am only interested in the cost calculation of the current state. what happened previously is a nice to have but most likely seldomly used.

**Comment by alikhajeh1** [comment]
# The ideas in this discussion went into action!

See https://www.infracost.io/docs/features/cli_commands/#comment-on-pull-requests for the comment options. 

[Thanks](https://www.infracost.io/blog/nov-2021-update/#community-highlights) to everyone who contributed to the discussion
